### PR TITLE
Sprint 51.2

### PR DESCRIPTION
**Pull request by ericsorenson**
# Sprint 51.2 

### In Progress

- [OPEN 6654 Performance Tuning - based on test legacy case migrations (Mike to QA Test)](https://github.com/flexion/ef-cms/issues/6654) 
- [OPEN 6684 Pending Report: Loading Issue](https://github.com/flexion/ef-cms/issues/6684) 
- [OPEN 6685 CMS User: Lambda timeout (Pick this story up first in the sprint)](https://github.com/flexion/ef-cms/issues/6685) 
- [OPEN 6833 Load Test in Prod: getCaseInventoryReport](https://github.com/flexion/ef-cms/issues/6833) 
- [OPEN 6836 Performance Tuning: Search Results (casePublicSearch, caseAdvancedSearch, advancedDocumentSearch & getPractitionersByName)](https://github.com/flexion/ef-cms/issues/6836) 
- [OPEN 6901 BUG: ServedPartiesCode of 'P' is not displaying correctly for migrated cases](https://github.com/flexion/ef-cms/issues/6901) 
- [OPEN 6908 BUG: Public Prod endpoints returning 503](https://github.com/flexion/ef-cms/issues/6908)

### CODE REVIEW

File: docs/architecture/email.md [hunk 1]
```diff
@@ -51,6 +51,19 @@ At present this is a manual process as a human needs to click on the link in the
...
+#### Automated
+
+1. Set the environment variables:
+  - `EFCMS_DOMAIN` e.g. `exp2.ustc-case-mgmt.flexion.us`
+  - `REGION` e.g. `us-east-1`
+  - `AWS_ACCESS_KEY_ID`
+  - `AWS_ACCOUNT_ID`
+  - `AWS_SECRET_ACCESS_KEY`
+2. Run the script:
+  - `cd web-api && ./verify-ses-email.sh`
```
Comment (mmarcotte) [inline docs/architecture/email.md:64]: 💪 Thanks @sutt0n !

File: shared/src/business/useCaseHelper/pendingItems/fetchPendingItems.js [hunk 2]
```diff
@@ -11,69 +7,30 @@ const { omit, pick } = require('lodash');
...
-  foundCases.forEach(foundCase => {
-    const foundCaseEntity = new Case(foundCase, {
+  const pendingItemResults = await applicationContext
+    .getPersistenceGateway()
+    .fetchPendingItems({
       applicationContext,
+      judge,
+      page,
+      source,
     });
```
Comment (mmarcotte) [inline shared/src/business/useCaseHelper/pendingItems/fetchPendingItems.js:29]: Nice bit of tidying!

File: shared/src/business/useCaseHelper/pendingItems/fetchPendingItems.test.js [hunk 3]
```diff
@@ -2,62 +2,34 @@ const {
...
-      .fetchPendingItems.mockReturnValue([mockDataOne, mockDataTwo]);
-    applicationContext
-      .getPersistenceGateway()
-      .getCaseByDocketNumber.mockReturnValue([mockDataOne, mockDataTwo]);
+      .fetchPendingItems.mockReturnValue(mockData);
   });
 
   it('calls search function with correct params when provided a judge and returns records', async () => {
-    const results = await fetchPendingItems({
+    const { foundDocuments } = await fetchPendingItems({
```
Comment (mmarcotte) [inline shared/src/business/useCaseHelper/pendingItems/fetchPendingItems.test.js]: I notice we introduced `total` as something we're returning from `fetchPendingItems`. Seems wise to test for it while we're testing for the foundDocuments as well -- unless there a reason why we're not testing it?

File: shared/src/business/useCaseHelper/pendingItems/fetchPendingItemsByDocketNumber.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  docketNumber,
+}) => {
+  const caseResult = await applicationContext
+    .getPersistenceGateway()
+    .getCaseByDocketNumber({
+      applicationContext,
+      docketNumber,
+    });
+
+  const caseEntity = new Case(caseResult, { applicationContext });
```
Comment (mmarcotte) [inline shared/src/business/useCaseHelper/pendingItems/fetchPendingItemsByDocketNumber.js]: seems like a good place to deconstruct `caseEntity` into
- docketEntries
- associatedJudge
- caseCaption
- docketNumber
- docketNumberSuffix
- status

File: shared/src/business/useCases/pendingItems/fetchPendingItemsInteractor.js [hunk 5]
```diff
@@ -12,14 +12,22 @@ const { UnauthorizedError } = require('../../../errors/errors');
...
+  page,
+}) => {
   const authorizedUser = applicationContext.getCurrentUser();
 
   if (!isAuthorized(authorizedUser, ROLE_PERMISSIONS.PENDING_ITEMS)) {
     throw new UnauthorizedError('Unauthorized');
   }
 
+  if (!judge) {
+    throw new Error('judge is required');
```
Comment (mmarcotte) [inline shared/src/business/useCases/pendingItems/fetchPendingItemsInteractor.js:28]: Woohoo another error. Can we write a quick test to make sure that we are throwing an error if we don't pass a judge :)

File: shared/src/persistence/elasticsearch/fetchPendingItems.js [hunk 6]
```diff
@@ -1,33 +1,69 @@
 const { search } = require('./searchClient');
 
-exports.fetchPendingItems = async ({ applicationContext, judge, source }) => {
+exports.fetchPendingItems = async ({
+  applicationContext,
+  judge,
+  page,
+  source,
+}) => {
+  const { PENDING_ITEMS_PAGE_SIZE } = applicationContext.getConstants();
```
Comment (mmarcotte) [inline shared/src/persistence/elasticsearch/fetchPendingItems.js:9]: Great to have this scaffolding in place.

File: shared/src/persistence/elasticsearch/fetchPendingItems.test.js [hunk 7]
```diff
@@ -8,20 +8,20 @@ const { search } = require('./searchClient');
 describe('fetchPendingItems', () => {
   it('returns results a query without judge', async () => {
     search.mockReturnValue({ results: ['some', 'matches'], total: 0 });
-    const results = await fetchPendingItems({
+    const { results } = await fetchPendingItems({
```
Comment (mmarcotte) [inline shared/src/persistence/elasticsearch/fetchPendingItems.test.js]: Again, we're including `total` in this return. Seems like we should include a quick test for it.

File: shared/src/proxies/pendingItems/fetchPendingItemsProxy.js [hunk 8]
```diff
@@ -8,12 +8,13 @@ const { get } = require('../requests');
...
-  if (judge) {
-    queryString = '?judge=' + encodeURIComponent(judge);
-  }
+exports.fetchPendingItemsInteractor = ({
+  applicationContext,
+  judge,
+  page = 0,
+}) => {
+  let queryString = `?page=${page}`;
+  queryString += '&judge=' + encodeURIComponent(judge);
```
Comment (mmarcotte) [inline shared/src/proxies/pendingItems/fetchPendingItemsProxy.js:17]: Logged a refactor item to consider `URLSearchParams`

File: web-api/verify-ses-email.sh [hunk 9]
```diff
@@ -0,0 +1,48 @@
+#!/bin/bash -e
```
Comment (mmarcotte) [inline web-api/verify-ses-email.sh:1]: thanks once again @sutt0n 😅

File: web-client/src/presenter/presenter-public.js [hunk 10]
```diff
@@ -31,6 +35,13 @@ import { validateOpinionSearchSequence } from './sequences/validateOpinionSearch
 import { validateOrderSearchSequence } from './sequences/validateOrderSearchSequence';
 
 export const presenter = {
+  catch: [
+    // ORDER MATTERS! Based on inheritance, the first match will be used
+    [InvalidRequestError, setCurrentPageErrorSequence], // 418, other unknown 4xx series
+    [ServerInvalidResponseError, setCurrentPageErrorSequence], // 501, 503, etc
+    [NotFoundError, notFoundErrorSequence], //404
+    [ActionError, setCurrentPageErrorSequence], // generic error handler
+  ],
```
Comment (mmarcotte) [inline web-client/src/presenter/presenter-public.js:44]: really nice to have!

File: shared/src/business/useCases/pendingItems/fetchPendingItemsInteractor.js [hunk 11]
```diff
@@ -12,14 +12,22 @@ const { UnauthorizedError } = require('../../../errors/errors');
  * @param {string} providers.judge the optional judge filter
```
Comment (julialeague) [inline shared/src/business/useCases/pendingItems/fetchPendingItemsInteractor.js:12]: The jsdoc needs to be updated to include "page", looks like

File: shared/src/business/useCases/pendingItems/generatePrintablePendingReportInteractor.js [hunk 12]
```diff
@@ -29,11 +29,21 @@ exports.generatePrintablePendingReportInteractor = async ({
...
-  const formattedPendingItems = pendingItems.map(pendingItem => ({
+  if (docketNumber) {
+    pendingDocuments = await applicationContext
+      .getUseCaseHelpers()
+      .fetchPendingItemsByDocketNumber({ applicationContext, docketNumber });
+  } else {
+    pendingDocuments = (
+      await applicationContext
+        .getUseCaseHelpers()
+        .fetchPendingItems({ applicationContext, judge })
```
Comment (julialeague) [inline shared/src/business/useCases/pendingItems/generatePrintablePendingReportInteractor.js:42]: I'm ignorant about interactors, but would you be able to use the fetchPendingItemsInteractor here?

**Comment by julialeague** [comment]
- [x] review the pull request itself, to get oriented
  - [x] read the description of the pull request, which should summarize the changes made
  - [x] read through every task on the Scrum board that's encompassed by this pull request
  - [x] read the description of the commits that comprise the pull request
- [x] If changes were made in the UI (if not, skip):
  - [x] If desired, run `docker system prune` to remove any unused docker images from previous code review.
  - [x] Fetch the pull request for the sprint (e.g., `git fetch origin pull/{PR #}/head:sprint-{Sprint #}`), and then switch to that branch (e.g. `git checkout sprint-{Sprint #}`)
  - [x] stand up the site locally, with `./docker-run.sh`
	- [x] test all functionality in all major browsers, emphasizing the functionality that this pull request addresses
	- [x] for internal Court functionality, perform the most thorough testing in Chrome, though also test in Edge and Firefox
	- [x] for public-facing functionality, test in browsers consistent with [public browser use data](https://analytics.usa.gov/)
	- [x] test in Mobile Safari and Mobile Chrome (or an emulator like Chrome DevTools), with the caveat that not all internal Court functionality will be necessary on these platforms
	- [x] use an automated audit tool for code quality and practices (recommended: [Chrome DevTools](https://developers.google.com/web/tools/chrome-devtools/), aka [Lighthouse](https://chrome.google.com/webstore/detail/lighthouse/blipmdconlkpinefehnmjammfjpmpbjk))
		- [x] look at efficiency of page loads, asset sizes, HTTP connection management, etc.
	- [x] review for accessibility
		- [x] use an automated audit tool, such as Chrome Audit or aXe
		- [x] navigate site only with the keyboard
		- [x] use VoiceOver or Narrator to navigate the site with audio only, with the display turned off
		- [x] manually test anything that pa11y cannot test automatically (e.g., contrast of text over images)
- [x] [review static code analysis results](https://sonarcloud.io/organizations/flexion-github/projects) in the vendor’s SonarCloud environment
- [x] examine OWASP ZAP output in `docs/`, to ensure that any errors are known to be false positives or have been previously declared to be acceptable
- [x] skim all new code, in the context of existing code, [looking for problems](#what-we-look-for) (knowing that the vast majority of new code will be covered by tests)
- [x] review all tests
  - [x] look at code coverage of tests in [the vendor's SonarCloud environment](https://sonarcloud.io/organizations/flexion-github/projects)
  - [x] methodically review all new tests for correctness, quality of naming
- [x] determine what code isn’t tested, review that rigorously
- [x] review documentation to ensure that it matches changes
  - [ ] user-visible changes (including API users like the IRS) are documented in CHANGELOG.md (which is linked from US Tax Court website).
- [x] provide comments on the pull request on GitHub, as necessary
  - [x] for comments that are specific to a particular line of code, comment on those specific lines
- [x] for each feature-level bug (i.e., it’s working as designed, but designed wrong), open a new issue and put it in the backlog

File: shared/src/business/useCases/pendingItems/fetchPendingItemsInteractor.js (hunk 5 above)
Comment (ericsorenson) [reply shared/src/business/useCases/pendingItems/fetchPendingItemsInteractor.js:28]: Test coming!

File: shared/src/business/useCaseHelper/pendingItems/fetchPendingItemsByDocketNumber.js (hunk 4 above)
Comment (ericsorenson) [reply shared/src/business/useCaseHelper/pendingItems/fetchPendingItemsByDocketNumber.js]: Update coming...

File: shared/src/business/useCaseHelper/pendingItems/fetchPendingItems.test.js (hunk 3 above)
Comment (ericsorenson) [reply shared/src/business/useCaseHelper/pendingItems/fetchPendingItems.test.js]: Update coming!

File: shared/src/persistence/elasticsearch/fetchPendingItems.test.js (hunk 7 above)
Comment (ericsorenson) [reply shared/src/persistence/elasticsearch/fetchPendingItems.test.js]: Update coming!

File: shared/src/business/useCases/pendingItems/generatePrintablePendingReportInteractor.js (hunk 12 above)
Comment (ericsorenson) [reply shared/src/business/useCases/pendingItems/generatePrintablePendingReportInteractor.js:42]: _Interactor_ and _use case_ are synonymous in this framework. Ideally an interactor function is only used for a single use case. When multiple use cases have common business functions, we extract those to use case helpers.

File: shared/src/business/useCases/pendingItems/fetchPendingItemsInteractor.js (hunk 11 above)
Comment (ericsorenson) [reply shared/src/business/useCases/pendingItems/fetchPendingItemsInteractor.js:12]: Updated!

File: shared/src/persistence/elasticsearch/fetchPendingItems.test.js (hunk 7 above)
Comment (ericsorenson) [reply shared/src/persistence/elasticsearch/fetchPendingItems.test.js]: Updated!

File: shared/src/business/useCases/pendingItems/fetchPendingItemsInteractor.js (hunk 5 above)
Comment (ericsorenson) [reply shared/src/business/useCases/pendingItems/fetchPendingItemsInteractor.js:28]: Updated!

File: shared/src/business/useCaseHelper/pendingItems/fetchPendingItemsByDocketNumber.js (hunk 4 above)
Comment (ericsorenson) [reply shared/src/business/useCaseHelper/pendingItems/fetchPendingItemsByDocketNumber.js]: Updated!

File: shared/src/business/useCaseHelper/pendingItems/fetchPendingItems.test.js (hunk 3 above)
Comment (ericsorenson) [reply shared/src/business/useCaseHelper/pendingItems/fetchPendingItems.test.js]: Updated!

File: shared/src/business/useCases/pendingItems/fetchPendingItemsInteractor.js (hunk 5 above)
Comment (mmarcotte) [reply shared/src/business/useCases/pendingItems/fetchPendingItemsInteractor.js:28]: ty

**Review by julialeague — APPROVED** [review]
@mmarcotte I recommend you accept these changes into the court environments!

**Pull request merged**